VERSIONING.md

```diff
@@ -0,0 +1,244 @@
+functionalities are currently in the `Sdk` class of the main package. eg:
+`Sdk.CreateTracerProviderBuilder()`, `Sdk.CreateMeterProviderBuilder()` etc. If
+metrics are in a separate package, the `Sdk` class cannot provide these entry
+points until the metrics code is moved to the main package. So users who use the
+experimental packages, will have to do code changes, when they upgrade to the
+main package consisting of the metrics.
+
+### Example with Approach2
+
+Following shows an example on how the `OpenTelemetry` package versioning works
+with this approach:
+
+`OpenTelemetry` 0.7.0-beta1 release : Pre-release, no API guarantees.
+
+`OpenTelemetry` 1.0.0-RC1 release : Pre-release, no API guarantees, but more
+stable than beta.
+
+`OpenTelemetry` 1.0.0 release : Stable release consisting of only stable signals
+: Traces, Propagators, Baggage.
+
+`OpenTelemetry.Metrics` 1.0.0-alpha release : Pre-release consisting of Metric
+SDK. Alpha indicates early stages of development. Metrics entry point will be
+contained in this package. (eg: new MeterProviderBuild().Build())
+
+`OpenTelemetry` 1.0.1 release : Bug fixes.
+
+`OpenTelemetry` 1.1.0 release : New features added. No metric code yet.
+
+`OpenTelemetry.Metrics` 1.0.0-beta release : Metric evolves to beta status.
+Still a pre-release.
+
+`OpenTelemetry.Metrics` 1.0.0-RC release : Metric evolves to RC status. Still a
+pre-release, but a final RC release will be done when Metrics is declared
+stable.
+
+`OpenTelemetry` 1.2.0 release : Add metric support. This will be additive
+changes. Metrics entry point will be added the main package (just like Traces.)
+(eg: Sdk.CreateMeterProviderBuilder().Build()) User who were previously using
+the *final* RC package of `OpenTelemetry.Metrics`, will need to change the
+Metric entry point code. Apart from that, they'll simply remove
+`OpenTelemetry.Metrics` and update `OpenTelemetry` to 1.2.0, with no *other*
+code changes.
+
+In the case of `OpenTelemetry.API.Metrics` package, users will require code
+changes, as the plan for Metrics API is to have it come from the .NET Runtime.
+
+`OpenTelemetry` 1.3.0 release : More features.
+
+`OpenTelemetry` 2.0.0 release : Remove any Obsolete methods from 1.* releases
+
+### Approach 3 - Same package name, but version differently
+
+The approach keeps the same amount of packages, but release (multiple) versions
+of the same, with experimental signals part of non-stable versions.
+
+This involves managing more than one branch in Github - master for regular work,
+and a separate experimental branch for experimental features. There can be 'n'
+experimental branches, if there is a need. The immediate future will have
+metrics only as separate branch.
+
+Release stable features as stable packages (OpenTelemetry 1.0.0, 1.1.0 etc.),
+from master branch. Release experimental features as different versions of the
+same package (OpenTelemetry 1.5.0-alpha.1), from experimental branch.
+
+### Downsides with Approach 3
```

Comment:
PRs may be come a bit more complicated in the event that there are changes that are desirable in both master and experimental branches. Though maybe this will be a good thing since it will promote smaller PRs 😄.

Reply:
could you add an example to better understand this part/

Reply:
Good question. It's tough to come up with a concrete example, but this could potentially arise when working with any cross-cutting concern.

Here's a PR that may help #1428 - though admittedly may not be the best example.

In this PR a new feature was introduced to the propagation API. The PR also changes all the instrumentation projects to use this new feature.

My thought is that there may be some work required on a cross-cutting concern like propagation in order to enable some work on metrics. It may be desirable to follow a flow like this:

1. Do the work on the propagation API on the main branch.
2. Review, approve, merge the work.
3. Update the experimental branch from main.
4. Begin the work on the experimental branch that depends on the propagation API work.

I don't think this is terribly complicated, but it does require some additional overhead to decide how best to order and land the work.

Reply:
Thanks for clarifying more. Agree to the steps you mentioned. 
Step3 in your flow is something we need to do frequently to ensure metrics is bought up to date with master, as metrics branch contains not only metrics code, but the tracing/baggage code as well.

Reply:
This makes sense to me too!

Using this example I wanted to bring up another situation. So far we're assuming that all metrics work is purely additive and the required cross-cutting work (step 1) can in fact be added to the main branch without breaking API/behavior.

Could a situation arise where that would not be the case? What if there was some public API refactoring we really wanted to do, because it would make experimental work a lot easier, or help make performance on the experimental work significantly better?

It is a purely hypothetical situation, and I can't think of a good specific example, but maybe something like https://github.com/open-telemetry/opentelemetry-dotnet/pull/1328, where we introduced BaseProcessor and BaseExporter?

"Approach 3" would still work nicely in this case (compared to "Approach 2", I think), although it would mean the experimental work is 2.0.0-alpha, instead of 1.5.0-alpha of course.

Again, just a theoretical example and probably unlikely to happen 😄. If it did come up though, would our approach be to still _not_ do the breaking change, release experimental as 1.5.0, but leave the refactoring as a separate thing for 2.0.0 later?

Reply:
> experimental work is 2.0.0-alpha, instead of 1.5.0-alpha of course.

From Otel spec about versioning, and in the context of metrics being the experimental feature - we should do 1.5.* only, and *not* 2.0.0. (Spec suggests 2.0 when we are ready to cut off features, not when we want to add something. So adding metrics must be a 1.xx release)